hibernate-core/src/main/java/org/hibernate/internal/ScrollableResultsImpl.java

```diff
@@ -122,15 +123,23 @@ private void prepareCurrentRow(boolean underlyingScrollSuccessful) {
 			return;
 		}
 
-		currentRow = getRowReader().readRow(
-				getRowProcessingState(),
-				getProcessingOptions()
-		);
+		final PersistenceContext persistenceContext = getPersistenceContext().getPersistenceContext();
+
+		persistenceContext.beforeLoad();
+		try {
+			currentRow = getRowReader().readRow(
+					getRowProcessingState(),
+					getProcessingOptions()
+			);
 
-		getRowProcessingState().finishRowProcessing();
-		getJdbcValuesSourceProcessingState().finishUp();
+			getRowProcessingState().finishRowProcessing();
+			getJdbcValuesSourceProcessingState().finishUp();
 
-		getRowProcessingState().getSession().getPersistenceContext().initializeNonLazyCollections();
+			getRowProcessingState().getSession().getPersistenceContext().initializeNonLazyCollections();
+		}
+		finally {
+			persistenceContext.afterLoad();
```

Comment:
I think this must happen before calling `initializeNonLazyCollections`, otherwise the loadCounter will always be `!= 0` and we will never load lazy collections.

Reply:
good catch :+1: